src/browser/modules/D3Visualization/lib/visualization/components/graphGeometry.ts

```diff
@@ -50,118 +60,138 @@ export default class NeoD3Geometry {
+  const maxCaptionTextLength = Math.floor(
+    (Math.pow(node.radius, 2) * Math.PI) / Math.pow(fontSize, 2)
+  )
   const template = style.forNode(node).get('caption')
   const nodeText = style.interpolate(template, node)
   const captionText =
-    nodeText.length > 100 ? nodeText.substring(0, 100) : nodeText
-  const fontFamily = 'sans-serif'
-  const fontSize = parseFloat(style.forNode(node).get('font-size'))
-  const lineHeight = fontSize
-  const measure = (text: any) => measureText(text, fontFamily, fontSize)
+    nodeText.length > maxCaptionTextLength
+      ? nodeText.substring(0, maxCaptionTextLength)
+      : nodeText
+  const measure = (text: string) =>
+    measureText(text, fontFamily, fontSize, canvas2DContext)
+  const whiteSpaceMeasureWidth = measure(' ')
 
   const words = captionText.split(' ')
 
-  const emptyLine = function(lineCount: any, iLine: any) {
-    let baseline = (1 + iLine - lineCount / 2) * lineHeight
-    if (style.forNode(node).get('icon-code')) {
-      baseline = baseline + node.radius / 3
-    }
-    const containingHeight =
-      iLine < lineCount / 2 ? baseline - lineHeight : baseline
-    const lineWidth =
-      Math.sqrt(square(node.radius) - square(containingHeight)) * 2
+  const emptyLine = (lineCount: number, lineIndex: number): NodeCaptionLine => {
+    // Calculate baseline of the text
+    const baseline = (1 + lineIndex - lineCount / 2) * fontSize
+
+    // The furthest distance between chord (top or bottom of the line) and circle centre
+    const chordCentreDistance =
+      lineIndex < lineCount / 2
+        ? baseline - fontSize / 2
+        : baseline + fontSize / 2
+    const maxLineWidth =
+      Math.sqrt(Math.pow(node.radius, 2) - Math.pow(chordCentreDistance, 2)) * 2
     return {
       node,
       text: '',
       baseline,
-      remainingWidth: lineWidth
+      remainingWidth: maxLineWidth
     }
   }
 
-  const fitOnFixedNumberOfLines = function(lineCount: any): [any, number] {
+  const addShortenedNextWord = (
+    line: NodeCaptionLine,
+    word: string
+  ): string => {
+    while (word.length > 2) {
+      const newWord = `${word.substring(0, word.length - 2)}\u2026`
+      if (measure(newWord) < line.remainingWidth) {
+        return `${line.text
+          .split(' ')
+          .slice(0, -1)
+          .join(' ')} ${newWord}`
+      }
+      word = word.substring(0, word.length - 1)
+    }
+    return `${word}\u2026`
+  }
+
+  const fitOnFixedNumberOfLines = function(
+    lineCount: number
+  ): [NodeCaptionLine[], number] {
     const lines = []
-    let iWord = 0
-    for (
-      let iLine = 0, end = lineCount - 1, asc = end >= 0;
-      asc ? iLine <= end : iLine >= end;
-      asc ? iLine++ : iLine--
-    ) {
-      const line = emptyLine(lineCount, iLine)
+    const wordMeasureWidthList: number[] = []
+    words.forEach((word: string) =>
+      wordMeasureWidthList.push(measure(`${word}`))
+    )
```

Comment:
Doesn't really matter (don't need to change it if you don't feel like it), but is there any reason not write this as:

```typescript
const wordMeasureWidthList = words.map(measure)
```
or
```typescript
const wordMeasureWidthList = words.map(w => measure(w))
```